quarkchain/cluster/miner.py

```diff
@@ -227,51 +238,85 @@ def _mine_new_block_async(self):
                 return
 
             instance.process = AioProcess(
-                target=instance.mine_func,
-                args=(block, instance.input_q, instance.output_q, mining_params),
+                target=instance._mine_loop,
+                args=(
+                    instance.consensus_type,
+                    block,
+                    instance.input_q,
+                    instance.output_q,
+                    mining_params,
+                ),
             )
             instance.process.start()
             await handle_mined_block(instance)
 
+        async def loop_for_work_gen(instance: Miner):
```

Comment:
Why not create work on demand i.e. only when get_work is called?
The work can still be cached and only create new one if the latest one was created more than 5 seconds ago.
That way the logic can be driven by external miner rather than a separate `enabled` variable. If the external miner stopped asking for work it will stop creating new work. The current code has to be stopped by an explicitly `setMining=False` JRPC call.

I also feel these code doesn't necessarily belong to the Miner class because it's actually handling the server side logic for miner. If we separate them out we can drop the `only_remote` check and `enabled` check if we change it to demand.

Reply:
makes sense. the current logic sorta mimics go-ethereum, where the `current_work` is generated either when a block is added or every 5 sec, using an event subscription model.

purely driven by clients + caching should be fine. also should find a way to clear current work once a new block has been added, so clients won't receive outdated work.

will refactor the miner server logic to another file, like `remote_miner` or `miner_client`

Reply:
having a second thought: from master server and shard's point of view, whether the miner is actually a mining doesn't really matter, they merely provide two callbacks (`add` and `create`) to the miner class and kick off the mining process (and only pass `get_work` / `submit_work` calls to them if mining is done outside). I'll keep the current `Miner` class, but remove the update-work-every-5-second logic.